add nearest neighbor lookup with KDTree for point to point distances

cchwala opened the pull request:
This PR adds an easy to use function for point data xarray.Datasets to calculate nearest neighbor distances using `scipy.spatial.KDTree`. The results are returned in a nicely structured xarray.Dataset.

The example notebook is updated accordingly, also removing some of the usage of the distance matrix because getting nearest neighbors is more convenient with the introduced functions.


- [x] closes #37 
- [x] Tests added and passed if fixing a bug or adding a new feature
- [x] All code checks from pre-commit passed
- [x] Added an entry in the CHANGELOG.md file

cchwala commented:
IMO the solution I currently have for returning the distance and neighbor IDs is good. It is borrowed and modified from #43.

I am not sure if is good to have `np.nan` as entry in the array of `neighbor_id` because this results in a `dtype='object'` because strings and floats are mixed. The advantage of having `np.nan` there is that one can easily drop these entries in xarray via `xarray.DataArray..dropna()` e.g. as I do in the example notebook to get the neighbor IDs for a certain PWS which I can then use to select data or locations from all neighbors:
```
neighbor_ids = closest_neigbors.sel(id=pws_id).neighbor_id.dropna(dim="n_closest")
ds_pws_neighbors = ds_pws.sel(id=neighbor_ids)
```
The `.sel()` of course only works if there is no non-matching ID in `neighbor_ids`.

cchwala commented:
@eoydvin It would be good if you could have a look at my code and notebook, to check it, but also to see how it fits with your PR #43.

eoydvin commented:
> IMO the solution I currently have for returning the distance and neighbor IDs is good. It is borrowed and modified from #43.
> 
> I am not sure if is good to have `np.nan` as entry in the array of `neighbor_id` because this results in a `dtype='object'` because strings and floats are mixed. The advantage of having `np.nan` there is that one can easily drop these entries in xarray via `xarray.DataArray..dropna()` e.g. as I do in the example notebook to get the neighbor IDs for a certain PWS which I can then use to select data or locations from all neighbors:
> 
> ```
> neighbor_ids = closest_neigbors.sel(id=pws_id).neighbor_id.dropna(dim="n_closest")
> ds_pws_neighbors = ds_pws.sel(id=neighbor_ids)
> ```
> 
> The `.sel()` of course only works if there is no non-matching ID in `neighbor_ids`.



> IMO the solution I currently have for returning the distance and neighbor IDs is good. It is borrowed and modified from #43.
> 
> I am not sure if is good to have `np.nan` as entry in the array of `neighbor_id` because this results in a `dtype='object'` because strings and floats are mixed. The advantage of having `np.nan` there is that one can easily drop these entries in xarray via `xarray.DataArray..dropna()` e.g. as I do in the example notebook to get the neighbor IDs for a certain PWS which I can then use to select data or locations from all neighbors:
> 
> ```
> neighbor_ids = closest_neigbors.sel(id=pws_id).neighbor_id.dropna(dim="n_closest")
> ds_pws_neighbors = ds_pws.sel(id=neighbor_ids)
> ```
> 
> The `.sel()` of course only works if there is no non-matching ID in `neighbor_ids`.

Hm, good point. In the newest version of #43 I use None instead of np.nan when I first create the array using `np.full([ds_cmls.cml_id.size, n_closest], None)`. Im not sure what is the best way to do it, but maybe it is good if the two functions do it the same way?

eoydvin commented on src/poligrain/spatial.py:
```diff
@@ -86,17 +86,91 @@ def project_point_coordinates(
+    tree_neighbors = scipy.spatial.KDTree(
+        data=list(zip(x_neighbors.values, y_neighbors.values))
+    )
+    distances, ixs = tree_neighbors.query(
+        list(zip(x.values, y.values)),
+        k=n_closest,
+        distance_upper_bound=max_distance,
+    )
```
when passing n_closest=1 to the function this line retruns a 1D array instead of the required 2D array, causing an error

eoydvin replied:
I solved this for the 1D case (where passing n_closest = 1 returned an integer using np.atleast_1d(ind_nearest_gauges). there should be an 2D equivalent, or you will need to do a test and then reshape(-1, 1)

cchwala replied:
thanks. good catch. I will add a test and then solve probably using your solution provided in the comment below.

cchwala commented:
Conclusion from discussion with @eoydvin 
1. Returning and xr.Dataset with the current structure is good. He will do that in #43, too.
2. It is better to have `None` instead of `np.nan` as entry in `neighbor_id` for the cases where no neighbor is found within `max_distance`. This is preferred because it makes more sense from a semantic point of view, to say that there is none neighbor ID. Using `.dropna()` e.g. `closest_neigbors.neighbor_id.sel(id='ams12').dropna(dim='n_closest')` still works with the entries being `None`.
3. We do not want to have all `coords` from `ds_points` in the returned dataset, only the `id` coordinate.

cchwala commented:
@eoydvin I implemented the points 2. and 3. from above. I also cleaned up the code and the notebook, which I also extended a bit. Please have a look again at this PR.

eoydvin reviewed (approved):
I think this is good!
